app/services/interface_cleaner.rb

```diff
@@ -0,0 +1,49 @@
+class InterfaceCleaner
+  delegate :logger, :to => :Rails
+
+  attr_reader :primary_nics, :provision_nics, :primary_hosts, :provision_hosts
+
+  def clean!
+    @primary_nics = []
+    @provision_nics = []
+    @primary_hosts = []
+    @provision_hosts = []
+    @deleted_count = delete_excluded_nics
+    self
+  end
+
+  def deleted_count
+    raise 'cleaner has not cleaned anything yet, run #clean! first' if @deleted_count.nil?
+    @deleted_count
+  end
+
+  private
+
+  def find_excluded_nics
+    logger.debug "Searching for nics that match excluded pattern..."
+    nic_ids = []
+
+    ignored_interfaces_regex = FactParser.ignored_interface_identifiers_regex
+
+    Nic::Base.unscoped.find_in_batches(:batch_size => 100) do |nics_batch|
```

Comment:
This will take a verrrrryyyy long time to iterate over all interfaces in a large set up. additionally, for every batch you iterate 5 more times on the matched interfaces to generate the id lists, which could also potentially require a large amount of memory.
The ignored syntax only allows `*` wildcard matching - which mean that you can search for matches using `LIKE` sql instead of matching a regex on each interface. How about something like:
```ruby
identifiers = Setting[:ignored_interface_identifiers].map{|i| i.tr('*','%')}
search = (["identifier LIKE ?"]*identifiers.length).join(" OR ")
Nic::Base.unscoped.where(search, *identifiers).select(:id, :host_id, :primary, :provision).find_each do |interface|
...
```

Reply:
I also think that affected systems has tens of millions of nic records and it's not good strategy to load them into memory this way. The key in being effective here is to do a single loop and delete them on the fly. That's basically what @tbrisker  suggest but I actually think that if we are going to filter them in one loop, it will actually be more efficient to do the filtering on the SQL-client side, because that where condition is for sure without index. Since we are talking about tens of millions of records, the query can eat significant amount of resources on the SQL-server side. Therefore I suggest to go even further to simple loop:

```ruby
Nic::Base.unscoped.select(:id, :host_id, :primary, :provision).find_each do |interface|
 # filter and delete them in this block - SQL server will do the transaction isolation for us it's safe to delete here
end

Reply:
Let me explain in more detail. Because we are using LIKE server cannot use indicies in this case. If we are iterating over simple table, table scan is actually not an issue if the query is unsorted for sure, but the problem here is at least one join. Optimizer can do weird things and can decide to join first, then we have huge problem. And I would not be surprised in bad optimization here because we will be submitting huge where clause here with many items. For this reason, I recommend to do simple loop to make sure this does not get de-optimized.

Iterating all records is sometimes faster and safer, I believe this is the case.

Reply:
I tested the query generated by my proposal on a DB with ~400k interfaces and 4k that match the patterns, and it ran in under one second:
```
foreman=# explain analyze SELECT  "nics"."id", "nics"."host_id", "nics"."primary", "nics"."provision" FROM "nics" WHERE (identifier LIKE 'lo' OR identifier LIKE 'usb%' OR identifier LIKE 'vnet%' OR identifier LIKE 'macvtap%' OR identifier LIKE '_vdsmdummy_' OR identifier LIKE 'veth%' OR identifier LIKE 'docker%' OR identifier LIKE 'tap%' OR identifier LIKE 'qbr%' OR identifier LIKE 'qvb%' OR identifier LIKE 'qvo%' OR identifier LIKE 'qr-%' OR identifier LIKE 'qg-%' OR identifier LIKE 'vlinuxbr%' OR identifier LIKE 'vovsbr%') ORDER BY "nics"."id" ASC LIMIT 1000;
                                                                                                                                                                           
             QUERY PLAN               
    
 Limit  (cost=9736.42..9737.99 rows=629 width=10) (actual time=258.907..259.025 rows=1000 loops=1)
   ->  Sort  (cost=9736.42..9737.99 rows=629 width=10) (actual time=258.905..258.974 rows=1000 loops=1)
         Sort Key: id
         Sort Method: top-N heapsort  Memory: 95kB
         ->  Seq Scan on nics  (cost=0.00..9707.18 rows=629 width=10) (actual time=0.038..257.565 rows=4208 loops=1)
               Filter: (((identifier)::text ~~ 'lo'::text) OR ((identifier)::text ~~ 'usb%'::text) OR ((identifier)::text ~~ 'vnet%'::text) OR ((identifier)::text ~~ 'macvtap%'::text) OR ((identifier)::text ~~ '_vdsmdummy_'::text) OR ((identifier)::text ~~ 'veth%'::text) OR ((identifier)::text ~~ 'docker%'::text) OR ((identifier)::text ~~ 'tap%'::text) OR ((identifier)::text ~~ 'qbr%'::text) OR ((identifier)::text ~~ 'qvb%'::text) OR ((identifier)::text ~~ 'qvo%'::text) OR ((identifier)::text ~~ 'qr-%'::text) OR ((identifier)::text ~~ 'qg-%'::text) OR ((identifier)::text ~~ 'vlinuxbr%'::text) OR ((identifier)::text ~~ 'vovsbr%'::text))
               Rows Removed by Filter: 373917
 Planning time: 0.120 ms
 Execution time: 259.064 ms
```
I think this is much better then instantiating 400k activerecord objects just to filter them out.

Reply:
That sort/order by must be removed. That's what took most of your time here :-) Anyway, can you test with find_each just to be sure Rails does not do some terrible magic trick there with WHERE 1=1 or anything like that. Just batch print them.

Reply:
the `sort` is caused by the `find_where` afaik, it needs to make sure it gets the batches in the correct order

Reply:
well actually, it uses the id sort for further filtering on next batches:
```
Nic::Base Load (36.5ms)  SELECT  "nics"."id", "nics"."host_id", "nics"."primary", "nics"."provision" FROM "nics" WHERE (identifier LIKE 'lo' OR identifier LIKE 'usb%' OR identifier LIKE 'vnet%' OR identifier LIKE 'macvtap%' OR identifier LIKE '_vdsmdummy_' OR identifier LIKE 'veth%' OR identifier LIKE 'docker%' OR identifier LIKE 'tap%' OR identifier LIKE 'qbr%' OR identifier LIKE 'qvb%' OR identifier LIKE 'qvo%' OR identifier LIKE 'qr-%' OR identifier LIKE 'qg-%' OR identifier LIKE 'vlinuxbr%' OR identifier LIKE 'vovsbr%') AND ("nics"."id" > 323241) ORDER BY "nics"."id" ASC LIMIT $1  [["LIMIT", 1000]]
```

Reply:
Ok, I am sold!